VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

DomSyna opened the pull request:
In various places in the code the wrong fragment cache is being checked when local TSB is enabled

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

In various places in the code the wrong fragment cache is being checked
when local TSB is enabled

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

In various places in the code the wrong fragment cache is being checked
when local TSB is enabled

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

cpc005 commented on StreamAbstractionAAMP.h:
```diff
@@ -85,7 +85,7 @@ struct StreamInfo
 	BitrateChangeReason reason;     /**< Reason for bitrate change*/
 	std::string baseUrl;
-	StreamInfo():enabled(false),isIframeTrack(false),validity(false),codecs(),bandwidthBitsPerSecond(0),resolution(),reason(){};
+	StreamInfo():enabled(false),isIframeTrack(false),validity(false),codecs(),bandwidthBitsPerSecond(0),resolution(),reason(),baseUrl(){};
```
I thought std::string xxx; has its own constructor that initialises it to empty string hence so no need to initialise it again in the class/struct constructor

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

In various places in the code the wrong fragment cache is being checked
when local TSB is enabled

Fixed crash due due to memset

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

DomSyna committed:
VPLAY-9825 [ENT-4082] Wrong fragment cache being checked

When the initial change to use the single fragment cache was done, some
code was missed.

Signed-off-by: DomSyna <192202460+DomSyna@users.noreply.github.com>

jfagunde reviewed (commented):
Can the changes made be tested in L1?

jfagunde commented on fragmentcollector_mpd.cpp:
```diff
@@ -8983,15 +8982,20 @@ void StreamAbstractionAAMP_MPD::AdvanceTrack(int trackIdx, bool trickPlay, doubl
-			if((pMediaStreamContext->numberOfFragmentsCached != maxCachedFragmentsPerTrack) && !(pMediaStreamContext->profileChanged) &&
-				(aamp->IsLocalAAMPTsbInjection() || !lowLatency || aamp->TrackDownloadsAreEnabled(static_cast<AampMediaType>(trackIdx))))
+
+			if(profileNotChanged && (isTsbInjection || (cacheNotFull && (!lowLatency || isTrackDownloadEnabled))))
```
How about testing this change in L1?
With the refactoring, the original code would be
`if(cacheNotFull && profiledNotChanged && (isTsbInjection || !lowLatency || isTrackDownloadEnabled))`

jfagunde commented on fragmentcollector_mpd.cpp:
```diff
@@ -9069,18 +9073,24 @@ void StreamAbstractionAAMP_MPD::AdvanceTrack(int trackIdx, bool trickPlay, doubl
 
-			if((pMediaStreamContext->numberOfFragmentsCached != maxCachedFragmentsPerTrack) && bCacheFullState &&
-				( !lowLatency || aamp->TrackDownloadsAreEnabled(static_cast<AampMediaType>(trackIdx))))
+			if ((isTsbInjection || (!pMediaStreamContext->IsFragmentCacheFull())) &&
```
How about testing this change in L1?
With the refactoring, the original code would be
`if(cacheNotFull && bCacheFullState && (!lowLatency || isTrackDownloadsEnabled))`

jfagunde commented on streamabstraction.cpp:
```diff
@@ -1829,27 +1855,46 @@ int MediaTrack::GetCurrentBandWidth()
 		{
-  			fragmentIdxToInject = 0;
+			AAMPLOG_DEBUG("[%s] Free cachedFragment[%d] numberOfFragmentsCached %d", name, fragmentIdxToInject, numberOfFragmentsCached);
+			mCachedFragment[fragmentIdxToInject].Clear();
```
How about fixing this under a separate Jira together with the code in MediaTrack::FlushFragments() and a L1 test to verify that Clear() method is called?

jfagunde reviewed (commented):
Looks good

pstroffolino committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-9825_FixForFragmentCacheChanges_1

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
819449774a4ad8ae48fe83fadd254f16
PRESENT
-->

# Coverity Issue - Data race condition
Accessing "pMediaStreamContext->numberOfFragmentChunksCached" without holding lock "MediaTrack.mutex". Elsewhere, "MediaTrack.numberOfFragmentChunksCached" is written to with "MediaTrack.mutex" held 3 out of 3 times (2 of these accesses strongly imply that it is necessary).

_Medium Impact, CWE-366_
_MISSING_LOCK_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[fragmentcollector_mpd.cpp:10102](undefined//blob/undefined/fragmentcollector_mpd.cpp#L10102)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
11f49e49df44bf5b9b57434e014fa506
PRESENT
-->

# Coverity Issue - Variable copied when it could be moved
"header" is passed-by-value as parameter to "std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""header"")" instead of "header".

## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[fragmentcollector_mpd.cpp:8781](undefined//blob/undefined/fragmentcollector_mpd.cpp#L8781)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
b7e9402d5325cfffffc304da46f07d5d
PRESENT
-->

# Coverity Issue - Double unlock
"PushNextFragment" unlocks "pMediaStreamContext->mutex" while it is unlocked.

_Medium Impact, CWE-765_
_LOCK_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[fragmentcollector_mpd.cpp:9039](undefined//blob/undefined/fragmentcollector_mpd.cpp#L9039)

rdkcmf-jenkins commented on streamabstraction.cpp:
```diff
@@ -1819,6 +1829,22 @@ CachedFragment* MediaTrack::GetFetchChunkBuffer(bool initialize)
+{
+	if(IsInjectionFromCachedFragmentChunks())
+	{
+		AAMPLOG_DEBUG("[%s] numberOfFragmentChunksCached %d mCachedFragmentChunksSize %zu", name, numberOfFragmentChunksCached, mCachedFragmentChunksSize);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
449fe91ba8e87de05bcf9e0fbc1c94b7
PRESENT
-->

# Coverity Issue - Data race condition
Accessing "this->numberOfFragmentChunksCached" without holding lock "MediaTrack.mutex". Elsewhere, "MediaTrack.numberOfFragmentChunksCached" is written to with "MediaTrack.mutex" held 3 out of 3 times (2 of these accesses strongly imply that it is necessary).

_Medium Impact, CWE-366_
_MISSING_LOCK_

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -9071,18 +9075,24 @@ void StreamAbstractionAAMP_MPD::AdvanceTrack(int trackIdx, bool trickPlay, doubl
-		AAMPLOG_ERR("%s Live downloader is not advancing at the moment cache (%d / %d)", GetMediaTypeName((AampMediaType) trackIdx), pMediaStreamContext->numberOfFragmentsCached, maxCachedFragmentsPerTrack);
+		if (pMediaStreamContext->IsInjectionFromCachedFragmentChunks())
+		{
+			AAMPLOG_ERR("%s Live downloader is not advancing at the moment cache (%d / %d)", GetMediaTypeName((AampMediaType)trackIdx), pMediaStreamContext->numberOfFragmentChunksCached, pMediaStreamContext->maxCachedFragmentChunksPerTrack);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0449a0ba518cd72585823907892ca556
PRESENT
-->

# Coverity Issue - Data race condition
Accessing "pMediaStreamContext->numberOfFragmentChunksCached" without holding lock "MediaTrack.mutex". Elsewhere, "MediaTrack.numberOfFragmentChunksCached" is written to with "MediaTrack.mutex" held 3 out of 3 times (2 of these accesses strongly imply that it is necessary).

_Medium Impact, CWE-366_
_MISSING_LOCK_
